//===----------------------------------------------------------------------===//
//
// This source file is part of the Teco open source project.
//
// Copyright (c) 2022 the Teco project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

// THIS FILE IS AUTOMATICALLY GENERATED by TecoServiceGenerator.
// DO NOT EDIT.

extension TCTmtError {
    public struct InternalError: TCErrorType {
        enum Code: String {
            case backendTimeout = "InternalError.BackendTimeout"
            case errorUnknown = "InternalError.ErrorUnknown"
            case requestFailed = "InternalError.RequestFailed"
            case other = "InternalError"
        }
        
        private let error: Code
        
        public let context: TCErrorContext?
        
        public var errorCode: String {
            self.error.rawValue
        }
        
        /// Initializer used by ``TCClient`` to match an error of this type.
        ///
        /// You should not use this initializer directly as there are no public initializers for ``TCErrorContext``.
        public init ?(errorCode: String, context: TCErrorContext) {
            guard let error = Code(rawValue: errorCode) else {
                return nil
            }
            self.error = error
            self.context = context
        }
        
        internal init (_ error: Code, context: TCErrorContext? = nil) {
            self.error = error
            self.context = context
        }
        
        /// 后台服务超时，请稍后重试。
        public static var backendTimeout: InternalError {
            InternalError(.backendTimeout)
        }
        
        /// 未知错误。
        public static var errorUnknown: InternalError {
            InternalError(.errorUnknown)
        }
        
        /// 请求失败。
        public static var requestFailed: InternalError {
            InternalError(.requestFailed)
        }
        
        /// 内部错误。
        public static var other: InternalError {
            InternalError(.other)
        }
    }
}

extension TCTmtError.InternalError: Equatable {
    public static func == (lhs: TCTmtError.InternalError, rhs: TCTmtError.InternalError) -> Bool {
        lhs.error == rhs.error
    }
}

extension TCTmtError.InternalError: CustomStringConvertible {
    public var description: String {
        return "\(self.error.rawValue): \(message ?? "")"
    }
}

extension TCTmtError.InternalError {
    /// - Returns: ``TCTmtError`` that holds the same error and context.
    public func toTmtError() -> TCTmtError {
        guard let code = TCTmtError.Code(rawValue: self.error.rawValue) else {
            fatalError("Unexpected internal conversion error!\nPlease file a bug at https://github.com/teco-project/teco to help address the problem.")
        }
        return TCTmtError(code, context: self.context)
    }
}

extension TCTmtError.InternalError {
    /// - Returns: ``TCCommonError`` that holds the same error and context.
    public func toCommonError() -> TCCommonError? {
        if let context = self.context, let error = TCCommonError(errorCode: self.error.rawValue, context: context) {
            return error
        }
        return nil
    }
}
